src/hugr/rewrite/consts.rs

```diff
@@ -0,0 +1,231 @@
+
+        Ok(source)
+    }
+
+    fn invalidation_set(&self) -> Self::InvalidationSet<'_> {
+        iter::once(self.0)
+    }
+}
+
+/// Remove a [`crate::ops::Const`] node with no outputs.
+#[derive(Debug, Clone)]
+pub struct RemoveConst(pub Node);
+
+/// Error from an [`RemoveConst`] operation.
+#[derive(Debug, Clone, Error, PartialEq, Eq)]
+pub enum RemoveConstError {
+    /// Invalid node.
+    #[error("Node is invalid (either not in HUGR or not Const).")]
+    InvalidNode(Node),
+    /// Node in use.
+    #[error("Node: {0:?} has non-zero outgoing connections.")]
+    ValueUsed(Node),
+    /// Removal error
+    #[error("Removing node caused error: {0:?}.")]
+    RemoveFail(#[from] HugrError),
+}
+
+impl Rewrite for RemoveConst {
+    type Error = RemoveConstError;
+
+    // The parent of the Const node.
+    type ApplyResult = Node;
+
+    type InvalidationSet<'a> = iter::Once<Node>;
+
+    const UNCHANGED_ON_FAILURE: bool = true;
+
+    fn verify(&self, h: &impl HugrView) -> Result<(), Self::Error> {
+        let node = self.0;
+
+        if (!h.contains_node(node)) || (!h.get_optype(node).is_const()) {
+            return Err(RemoveConstError::InvalidNode(node));
+        }
+
+        if h.output_neighbours(node).next().is_some() {
+            return Err(RemoveConstError::ValueUsed(node));
+        }
+
+        Ok(())
+    }
+
+    fn apply(self, h: &mut impl HugrMut) -> Result<Self::ApplyResult, Self::Error> {
+        self.verify(h)?;
+        let node = self.0;
+        let source = h
+            .get_parent(node)
+            .expect("Const node without a parent shouldn't happen.");
+        h.remove_node(node)?;
+
+        Ok(source)
+    }
+
+    fn invalidation_set(&self) -> Self::InvalidationSet<'_> {
+        iter::once(self.0)
+    }
+}
+
+#[rustversion::since(1.75)] // uses impl in return position
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::{
+        builder::{Container, Dataflow, HugrBuilder, ModuleBuilder, SubContainer},
+        extension::{
+            prelude::{ConstUsize, USIZE_T},
+            PRELUDE_REGISTRY,
+        },
+        hugr::HugrMut,
+        ops::{handle::NodeHandle, LeafOp},
+        type_row,
+        types::FunctionType,
+    };
+    #[test]
+    fn test_const_remove() -> Result<(), Box<dyn std::error::Error>> {
+        let mut build = ModuleBuilder::new();
+        let con_node = build.add_constant(ConstUsize::new(2))?;
+
+        let mut dfg_build =
+            build.define_function("main", FunctionType::new_endo(type_row![]).into())?;
+        let load_1 = dfg_build.load_const(&con_node)?;
+        let load_2 = dfg_build.load_const(&con_node)?;
+        let tup = dfg_build.add_dataflow_op(
+            LeafOp::MakeTuple {
+                tys: type_row![USIZE_T, USIZE_T],
+            },
+            [load_1, load_2],
+        )?;
+        dfg_build.finish_sub_container()?;
+
+        let mut h = build.finish_prelude_hugr()?;
+        assert_eq!(h.node_count(), 8);
```

Comment:
You could change this to a Function-rooted Hugr to reduce counts by 1. Would be nice to comment why there are so many (Module, Function, Input, Output, Const, LoadConstant*2, Tuple)...

Reply:
Trying to test that non-local const->load edges work ok
Will add comment